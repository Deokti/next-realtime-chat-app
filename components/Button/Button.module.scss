.button:hover {
  opacity: .9;
}

.button:active {
  transform: translateY(2px);
}

.active > * {
  display: flex;
  justify-content: center;
  align-items: center;
}

.isLoading,
.disabled {
  pointer-events: none;
  opacity: .6;
}


