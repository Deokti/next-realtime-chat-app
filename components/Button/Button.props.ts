import { ButtonHTMLAttributes, DetailedHTMLProps, ReactNode } from "react";

export interface ButtonProps extends DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  children: ReactNode
  apperance?: 'green' | 'red' | 'transparent'
  disabled?: boolean
  size?: 'large' | 'default' | 'full'
  loading: boolean
}