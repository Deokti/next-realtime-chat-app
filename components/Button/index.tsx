import styles from './Button.module.scss';

type TButton = {
  children: string
  width?: number | string
  height?: number | string
  border?: string
  cursor?: 'pointer' | 'text' | 'auto'
  backgroundColor?: string
  color?: string
  fontSize?: number
  fontWeight?: 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900 | number
  borderRadius?: number
  onClick?: () => {}
  type: 'button' | 'submit',
  isLoading?: boolean
  LoadingIcon?: JSX.Element | null | Function
}

function Button({
  type = 'button',
  children = 'Нажми на меня',
  width,
  height,
  border = 'none',
  cursor = 'pointer',
  backgroundColor,
  color = "#fff",
  fontSize = 16,
  fontWeight,
  borderRadius = 0,
  onClick,
  LoadingIcon,
  isLoading = false,
}: TButton) {

  const getLoadingIconByType = LoadingIcon === Function ? <LoadingIcon /> : LoadingIcon;

  return (
    <button
      className={`${styles.button} ${isLoading && styles.isLoading}`}
      type={type}
      style={
        {
          cursor,
          width,
          height,
          border,
          backgroundColor,
          color,
          fontSize,
          fontWeight,
          borderRadius
        }
      }
      disabled={isLoading}
      onClick={onClick}
    >
      {isLoading ? getLoadingIconByType : children}
    </button>
  );
}

export default Button;
