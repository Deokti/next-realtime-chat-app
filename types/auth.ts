export interface IUser {
  id: string
  avatar: string
  username: string
  isOnline: boolean
}

export interface ILogin {
  email: string
  password: string
}

export interface IRegister {
  username: string
  email: string
  password: string
  passwordConfirm: string
}

export interface IAuthError {
  a: any
  code: string
  message: string
}