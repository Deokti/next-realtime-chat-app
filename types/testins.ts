import { ShallowWrapper } from "enzyme";

export type TComponentTest = ShallowWrapper<any, Readonly<{}> | React.Component<{}, {}, any>>