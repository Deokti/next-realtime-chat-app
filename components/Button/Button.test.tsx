import React from 'react';
import Button, { ButtonProps } from './index';
import { shallow, ShallowWrapper } from 'enzyme';
import { TComponentTest } from '../../types/testins';

const button = (props?: ButtonProps) => shallow(<Button {...props}>{props?.children}</Button>)

const props: ButtonProps = {
  children: 'Тестовая кнопка',
  width: 250,
  height: 45,
  border: '1px solid #000',
  cursor: 'pointer',
  backgroundColor: '#fcfc',
  color: '#000',
  fontSize: 14,
  fontWeight: 500,
  borderRadius: 4,
  type: 'button',
  isLoading: false,
  disabled: false,
}

describe('Button component', () => {
  let component: TComponentTest
  beforeEach(() => component = button());

  it('should Button component containg one button tag', () => {
    expect(component.find('button').length).toEqual(1);
  });

  it('should Button component contais text - Аутентификация', () => {
    component = button({ children: 'Аутентификация' });
    expect(component.find('button').text()).toEqual('Аутентификация');
  });

  it('should render Button component', () => {
    component = button(props);
    expect(component).toMatchSnapshot();
  });

  it('should Button contain isLoading class if transferred isLoading prop', () => {
    component = button({ isLoading: true });
    expect(component.find('button.isLoading').length).toEqual(1);
  });

  it('should Button contain disabled class if transferred disabled prop', () => {
    component = button({ disabled: true });
    expect(component.find('button.disabled').length).toEqual(1);
  });
});