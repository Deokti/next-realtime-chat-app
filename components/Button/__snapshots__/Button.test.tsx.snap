// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render Button component 1`] = `
<button
  className="button"
  disabled={false}
  style={
    Object {
      "backgroundColor": "#fcfc",
      "border": "1px solid #000",
      "borderRadius": 4,
      "color": "#000",
      "cursor": "pointer",
      "fontSize": 14,
      "fontWeight": 500,
      "height": 45,
      "width": 250,
    }
  }
  type="button"
>
  Тестовая кнопка
</button>
`;
